cli/lsp/repl.rs

```diff
@@ -0,0 +1,292 @@
+      .initialize(InitializeParams {
+        process_id: None,
+        root_path: None,
+        root_uri: Some(cwd_uri.clone()),
+        initialization_options: Some(
+          serde_json::to_value(get_repl_workspace_settings()).unwrap(),
+        ),
+        capabilities: ClientCapabilities {
+          workspace: None,
+          text_document: None,
+          window: None,
+          general: None,
+          experimental: None,
+        },
+        trace: None,
+        workspace_folders: None,
+        client_info: Some(ClientInfo {
+          name: "Deno REPL".to_string(),
+          version: None,
+        }),
+        locale: None,
+      })
+      .await?;
+
+    language_server.initialized(InitializedParams {}).await;
+
+    let server = ReplLanguageServer {
+      language_server,
+      document_version: 0,
+      document_text: String::new(),
+      pending_text: String::new(),
+      cwd_uri,
+    };
+    server.open_current_document().await;
+
+    Ok(server)
+  }
+
+  pub async fn commit_text(&mut self, line_text: &str) {
+    self.did_change(line_text).await;
+    self.document_text.push_str(&self.pending_text);
+    self.pending_text = String::new();
+  }
+
+  pub async fn completions(
+    &mut self,
+    line_text: &str,
+    position: usize,
+  ) -> Vec<ReplCompletionItem> {
+    self.did_change(line_text).await;
+    let before_line_len = BytePos(self.document_text.len() as u32);
+    let position = before_line_len + BytePos(position as u32);
+    let text_info = deno_ast::SourceTextInfo::from_string(format!(
+      "{}{}",
+      self.document_text, self.pending_text
+    ));
+    let line_and_column = text_info.line_and_column_index(position);
+    let response = self
+      .language_server
+      .completion(CompletionParams {
+        text_document_position: TextDocumentPositionParams {
+          text_document: TextDocumentIdentifier {
+            uri: self.get_document_specifier(),
+          },
+          position: Position {
+            line: line_and_column.line_index as u32,
+            character: line_and_column.column_index as u32,
+          },
+        },
+        work_done_progress_params: WorkDoneProgressParams {
+          work_done_token: None,
+        },
+        partial_result_params: PartialResultParams {
+          partial_result_token: None,
+        },
+        context: None,
+      })
+      .await
+      .ok()
+      .unwrap_or_default();
+
+    let items = match response {
+      Some(CompletionResponse::Array(items)) => items,
+      Some(CompletionResponse::List(list)) => list.items,
+      None => Vec::new(),
+    };
+    items
+      .into_iter()
+      .filter_map(|item| {
+        item.text_edit.and_then(|edit| match edit {
+          CompletionTextEdit::Edit(edit) => Some(ReplCompletionItem {
+            new_text: edit.new_text,
+            span: lsp_range_to_span(&text_info, &edit.range),
+          }),
+          CompletionTextEdit::InsertAndReplace(_) => None,
```

Comment:
discussion: while we don't issue registry completions with insert and replace, it doesn't mean we _can't_ in the future, and we should potentially consider handling it here.  Certainly some that come back from tsc around other completions do use this.

Reply:
Yeah, I think it will be useful when implementing the other kinds of completions.